[ruijietay] iP

ruijietay opened the pull request:
# Capone

> _"He who fails to plan, plans to fail"_ - [_Winston Churchill_](https://en.wikipedia.org/wiki/Winston_Churchill)

**Capone is a lightweight and free task-tracking application - It's the only task tracking application you need.**

## Why Capone?
- It's free
- It's [source code](#Public-Source-Code) is public!
- It's [lightweight and portable](#Lightweight-and-portable)!
- It's [privacy-respecting](#Privacy-respecting)

### Public Source Code
Capone is based off the original greenfield project [Duke](https://github.com/nus-cs2103-AY2324S2/ip) that is part of NUS Computer Science Software Engineering Course. The source code for Capone is available [here](https://github.com/ruijietay/ip).

### Lightweight and Portable
With the entire application packaged into a single JAR file that is less than 1MB<sup>_(as of [v0.1](https://github.com/ruijietay/ip/releases/tag/A-Jar)_)</sup>, Capone is one of the most lightweight task tracking applications out in the market.

The tasks are stored in a single `JSON` file stored locally which can be synced using your favorite cloud provider.

### Privacy-respecting
Capone works totally offline - all tasks are stored in a single JSON file. Hence, no data is being collected. Just you and your tasks, forever :smiley:

## Installation
### Prerequisites
Capone is written in Java. Hence, please ensure that you have Java running on your device. You can download Java [here](https://www.java.com/en/download/).

### Downloading and Running Capone

#### Via JAR File
To use Capone, you only need 2 simple steps:
1. Download the latest version of [Capone](https://github.com/ruijietay/ip/releases) (JAR file).
2. Run the JAR file using the command `java -jar ./capone.jar`

#### From Source
Alternatively, if you are a programmer or you prefer to run Capone from source via an IDE such as IntellJ, you can find the main Capone method in `Capone.java`.

```java
    /**
     * The main method to start the Capone application.
     *
     * @param args Command-line arguments (not used).
     */
    public static void main(String[] args) {
        new Capone("./data/", "tasks.json").run();
    }
```



## Features
Support for various types of tasks
1. Todo
1. Deadline
1. Event

You can also interact with the tasks in a variety of ways:

- [x] Mark tasks as completed
- [ ] Unmark tasks
- Delete tasks
- Find tasks

> [!TIP]
> You can also use the `help` command to find out the full list of commands Capone supports.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ruijietay committed:
Given new name to chatbot.

ruijietay committed:
Refactor filenames to correct chatbot name.

ruijietay committed:
Implement echo and exit functionality.

ruijietay committed:
Implement ability to store task and list tasks.

ruijietay committed:
Changed primitive array to ArrayList for easier management. Added Task class.

ruijietay committed:
Implement the ability to mark and unmark tasks as done.

ruijietay committed:
Implement ToDo.

ruijietay committed:
Implement Deadline.

ruijietay committed:
Implement Event.

ruijietay committed:
Update README.md

ruijietay committed:
Update README.md

ruijietay committed:
Automated Text UI Testing

ruijietay committed:
Merge branch 'master' of github.com:ruijietay/ip

ruijietay committed:
Added decorative name

ruijietay committed:
Implement minimal exceptions

ruijietay committed:
Implmented more exception handling.

ruijietay committed:
Updated Automated Text UI Testing.

ruijietay committed:
Add support for deleting tasks from the list.

ruijietay committed:
Updated help menu.

ruijietay committed:
Updated Automated Text UI Testing.

ruijietay committed:
Used enums for commands for easier viewing of all commands.

ruijietay committed:
Implement saving of tasks to file

ruijietay committed:
Merge branch 'branch-Level-7'

ruijietay committed:
Date recognition for deadline tasks.

ruijietay committed:
Improved recognition for deadline tasks.

ruijietay committed:
Implement date and time recognition for tasks.

ruijietay committed:
Added test cases for date and time recognition.

ruijietay committed:
Merge branch 'branch-Level-8'

ruijietay committed:
Refactored code to be more Object-Oriented (OOP).

ruijietay committed:
Merge branch 'branch-A-MoreOOP'

ruijietay committed:
Refactor TaskStorage class to Storage.

ruijietay committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ruijietay committed:
Add gradle support

ruijietay committed:
Add JUnit tests

ruijietay committed:
Modified gradle config to create JAR file

ruijietay committed:
Merge branch 'branch-A-Jar'

ruijietay committed:
Added JavaDocs for all files.

ruijietay committed:
Comply with JavaDoc coding standard

ruijietay committed:
Implement a way to find a task by searching for a keyword.

ruijietay committed:
Merge branch 'branch-Level-9'

ruijietay committed:
Resolve merge conflict

ruijietay committed:
Resolve merge conflict

ruijietay committed:
Resolve merge conflict

ruijietay committed:
Resolve merge conflict

ruijietay committed:
Refactor code for better Ui separation between Cli and Gui

ruijietay committed:
Implement basic echo functionality of JavaFX GUI

ruijietay committed:
Implement GUI for Capone

ruijietay committed:
Merge branch 'branch-Level-10'

kab-dot commented on src/main/java/capone/Parser.java:
```diff
@@ -0,0 +1,169 @@
+     * @throws InvalidTimeException If the input string is not a valid time.
+     */
+    public static LocalTime parseTime(String time) throws InvalidTimeException {
+        DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HHmm");
```
Perhaps you could also include the input where the user types in HH:mm as that is also a common way of writing the time

ruijietay replied:
Yes, that's a good idea. I will consider your suggestion!

kab-dot commented on src/main/java/capone/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        // Combine description of task into one string.
+        StringBuilder description = new StringBuilder();
```
Maybe instead of using a loop, you could try using String.join() to concatenate the description parts from 1 to byNdx and String.substring() to extract the deadline part after /by

ruijietay replied:
Thank you! I will look into it 👍

kab-dot commented on src/main/java/capone/commands/ByeCommand.java:
```diff
@@ -0,0 +1,43 @@
+package capone.commands;
```
good that you separated the command into the different classes!

ruijietay committed:
Added javadocs and fixed checkstyle violations.

ruijietay committed:
Merge branch 'branch-A-Checkstyle'

kab-dot reviewed (commented):
Looks really good to go!

kailinteoo commented on src/main/java/capone/TaskList.java:
```diff
@@ -0,0 +1,94 @@
+ *
+ * @author Tay Rui-Jie
+ */
+public class TaskList implements Iterable<Task> {
```
Your TaskList arrangement is commendable, as it effectively separates commands, making it easier for future edits and enhancing readability for users. This thoughtful organization streamlines both code maintenance and comprehension.

kailinteoo commented on src/main/java/capone/commands/EventCommand.java:
```diff
@@ -0,0 +1,151 @@
+        // If /to is specified before /from, throw error.
+        if (toNdx < fromNdx) {
+            throw new InvalidCommandException("Please input from date followed by to date!\n"
+                    + "Usage: event [description] /from [date] /to [date]");
```
Your proactive approach in handling potential input errors without explicit instruction is commendable. It demonstrates foresight and consideration for user experience, ensuring smooth functionality even in unforeseen scenarios. Good job!

kailinteoo commented on src/main/java/capone/commands/EventCommand.java:
```diff
@@ -0,0 +1,151 @@
+ *
+ * @author Tay Rui-Jie
+ */
+public class EventCommand extends Command {
```
Your attention to detail and commitment to error management in this class really contribute to the reliability and robustness of your code!!

kailinteoo commented on src/main/java/capone/commands/InvalidCommand.java:
```diff
@@ -0,0 +1,30 @@
+    @Override
+    public String execute(TaskList taskList, Ui ui, Storage storage) throws InvalidCommandException {
+        throw new InvalidCommandException("I'm sorry, I don't understand what you just said.\n"
+                + "Use 'help' to display the list of valid commands");
```
The function displaying a list of valid commands is a clever addition, greatly aiding users by offering clear guidance and enhancing overall usability.

kailinteoo reviewed (commented):
Excellent work in maintaining consistency in error handling across the codebase! This consistency makes it easy for users to understand the flow of your code and how it operates. Additionally, your efforts to enhance readability and make the code more intuitive are commendable. Your use of Javadoc also provides clear explanations for classes and methods. Remember to keep the Javadoc updated as the code evolves and consider adding more detailed descriptions where necessary. Overall, your consistent code formatting makes the codebase more approachable. Keep up the great work!!

ruijietay committed:
Enable assertion for GUI and CLI

ruijietay committed:
Add assertions in various points of the code

Potential bugs in the code might be present due to assumptions made in
the code.

If such bugs are present in an production environement, it will cause
the program to break or throw an uncaught exception.

Assertions are used to to detect violations in assumptions we have made
in the code. For example, we assert that a newly initialised task list
file contains "[]".

Implementing such assertions allows the runtime to verify that the
assumtions we made holds.

Let's
* Assert that the task list is NOT empty before printing it out to the user
* Assert that a new task list file contains the "[]" characters
* Assert that the user should not have an valid dateTime entry when
  processing its dateTime entry as a String

ruijietay committed:
Add Javadoc to isEmpty method in TaskList

ruijietay committed:
Refactor Event class

Event class has some code duplication (e.g. when parsing LocalDateTimes)

These repetative code violates the DRY principle.

Extracting such repetative code into a static method under the Parser
class not only respects the DRY principle, but is also more object
oriented.

Let's refactor the parsing of the description and datetime objects into
a static method under the Parser class.

ruijietay committed:
Fix off-by-1 index bugs in Parser

ruijietay committed:
Refactor TodoCommand class

ruijietay committed:
Refactor DeadlineCommand class

ruijietay committed:
Implement streams in FindCommand

ruijietay committed:
Merge pull request #1 from ruijietay/branch-A-Assertions

 Add assertions in various points of the code

ruijietay committed:
Fix merge conflict

ruijietay committed:
Merge branch master

ruijietay committed:
Merge pull request #3 from ruijietay/branch-A-CodeQuality

Improve Code Quality - Refactor Event Class

ruijietay committed:
Merge branch 'master' into branch-A-CodeQuality

ruijietay committed:
Merge branch 'master' into branch-A-Streams

ruijietay committed:
Merge pull request #2 from ruijietay/branch-A-Streams

Implement streams in FindCommand

ruijietay committed:
Remove redundant code

ruijietay committed:
Improve GUI aesthetics

ruijietay committed:
Remove redundant code

ruijietay committed:
Implement update function to task description

ruijietay committed:
Merge pull request #4 from ruijietay/branch-BCD-Extension

Merge branch-BCD-Extension (C-Update)

ruijietay committed:
Fix minor checkstyle violations

ruijietay committed:
Add gradle.yml file for Github Actions

ruijietay committed:
Add previously untracked resource/config files

ruijietay committed:
Merge branch 'master' into branch-A-CodeQuality

ruijietay committed:
Refactor code to use constants

ruijietay committed:
Refactor code to reduce nested if-else statements

ruijietay committed:
Add previously untracked checkstyle config files

ruijietay committed:
Refactor code to better handle corrupted JSON file

ruijietay committed:
Merge pull request #5 from ruijietay/branch-A-CodeQuality

LGTM

ruijietay committed:
Merge branch 'master' of github.com:ruijietay/ip

ruijietay committed:
Fix minor checkstyle violations

ruijietay committed:
Fix minor checkstyle violation

ruijietay committed:
Improve GUI - add text prompt and change font

ruijietay committed:
Merge branch 'branch-A-BetterGui'

ruijietay committed:
Modify welcome message and chat bubble colour

ruijietay committed:
Merge branch 'branch-A-Personality'

ruijietay committed:
Add error handling for invalid start/end date pair

ruijietay committed:
Merge branch 'branch-A-MoreErrorHandling'

ruijietay committed:
Update DeadlineCommandTest for GUI testing

ruijietay committed:
Remove unused import statements

ruijietay committed:
Add testing for event commands

ruijietay committed:
Merge branch 'branch-A-MoreTesting'

ruijietay committed:
Fix minor checkstyle violations and JavaDocs

ruijietay committed:
Set title for Capone application

ruijietay committed:
Modify chatbox GUI

ruijietay committed:
Merge branch 'branch-A-BetterGUI'

ruijietay committed:
Fix test cases related to Event commands

ruijietay committed:
Add image for Capone UI

ruijietay committed:
Modify usage string for Deadline and Event command

ruijietay committed:
Modify usage string for Deadline and Event command

ruijietay committed:
Modify help command

ruijietay committed:
Update help command

ruijietay committed:
Update README.md for User Guide

ruijietay committed:
Add config file for Jekyll

ruijietay committed:
Update test cases

ruijietay committed:
Set wrap width for dialog box

ruijietay committed:
Update _config.yml

ruijietay committed:
Update _config.yml to change description
